File: 06/objc/vehicle.m
#import "vehicle.h"

@interface Vehicle()

@end

@implementation Vehicle

- (id) AmountOfDoor: (int) door AmountOfCylinder: (int) cylinder TransmissionType: (int) trans Color: (NSString *) color FuelLevel: (double) fuel {
	self = [super init];
	self.amountOfDoor = door;
	self.amountOfCylinder = cylinder;
	self.transmissionType = trans;
	self.color = color;
	self.fuelLevel = fuel;
	return self;
}

- (void) printDetail {
	printf("%s\n", [NSStringFromClass([self class]) UTF8String]);
	printf("\tNumber of Doors: %d\n", self.amountOfDoor);
	printf("\tNumber of Cylinders: %d\n", self.amountOfCylinder);
	printf("\tTransmission Type: %d\n", self.transmissionType);
	printf("\tColor: %s\n", [self.color UTF8String]);
	printf("\tFuel Level: %.2f\n", self.fuelLevel);
}

- (void) horn {
	printf ("Horn by the vehicle!\n");
}

@end

File: 06/vehicle.cpp
#include "vehicle.h"

using namespace std;

Vehicle::Vehicle(int amountOfDoor, int amountOfCylinder, int transmissionType, std::string color, double fuelLevel) {
	this->amountOfDoor_ = amountOfDoor;
	this->amountOfCylinder_ = amountOfCylinder;
	this->transmissionType_ = transmissionType;
	this->color_ = color;
	this->fuelLevel_ = fuelLevel;
}

void Vehicle::printDetail() {
	cout << typeid(*this).name() << endl;
	cout << "\tNumber of Doors: " << this->amountOfDoor_ << endl;
	cout << "\tNumber of Cylinders: " << this->amountOfCylinder_ << endl;
	cout << "\tTransmission Type: " << this->transmissionType_ << endl;
	cout << "\tColor: " << this->color_ << endl;
	cout << "\tFuel Level: " << this->fuelLevel_ << endl;
}

ostream &operator<<(ostream &out, const Vehicle &v) {
	out << typeid(v).name()
	<< "\n\tNumber of Doors: " << v.amountOfDoor_
	<< "\n\tNumber of Cylinders: " << v.amountOfCylinder_
	<< "\n\tTransmission Type: " << v.transmissionType_
	<< "\n\tColor: " << v.color_
	<< "\n\tFuel Level: " << v.fuelLevel_ << endl;
	return out;
}

Vehicle::~Vehicle() {
	//delete this;
}

void Vehicle::horn() {
	cout << "Horn by the vehicle!" << endl;
}
